core/src/ephemeral_server/mod.rs

```diff
@@ -0,0 +1,461 @@
+            args,
+            has_test_service: false,
+        })
+        .await
+    }
+}
+
+/// Configuration for the test server.
+#[derive(Debug, Clone, derive_builder::Builder)]
+pub struct TestServerConfig {
+    /// Required path to executable or download info.
+    pub exe: EphemeralExe,
+    /// Port to use or obtains a free one if none given.
+    #[builder(default)]
+    pub port: Option<u16>,
+    /// Additional arguments to the test server.
+    #[builder(default)]
+    pub extra_args: Vec<String>,
+}
+
+impl TestServerConfig {
+    /// Start a test server.
+    pub async fn start_server(&self) -> anyhow::Result<EphemeralServer> {
+        // Get exe path
+        let exe_path = self.exe.get_or_download("temporal-test-server").await?;
+
+        // Get free port if not already given
+        let port = self.port.unwrap_or_else(get_free_port);
+
+        // Build arg set
+        let mut args = vec![port.to_string()];
+        args.extend(self.extra_args.clone());
+
+        // Start
+        EphemeralServer::start(EphemeralServerConfig {
+            exe_path,
+            port,
+            args,
+            has_test_service: true,
+        })
+        .await
+    }
+}
+
+struct EphemeralServerConfig {
+    exe_path: PathBuf,
+    port: u16,
+    args: Vec<String>,
+    has_test_service: bool,
+}
+
+/// Server that will be stopped when dropped.
+pub struct EphemeralServer {
+    /// gRPC target host:port for the server frontend.
+    pub target: String,
+    /// Whether the target implements the gRPC TestService
+    pub has_test_service: bool,
+    child: tokio::process::Child,
+}
+
+impl EphemeralServer {
+    async fn start(config: EphemeralServerConfig) -> anyhow::Result<EphemeralServer> {
+        // Start process
+        // TODO(cretz): Offer stdio suppression?
+        let child = tokio::process::Command::new(config.exe_path)
+            .args(config.args)
+            // We kill this on drop in case of errors or other issues. We still
+            // encourage users to manually shutdown.
+            .kill_on_drop(true)
+            .spawn()?;
+        let target = format!("127.0.0.1:{}", config.port);
+        let target_url = format!("http://{}", target);
+        let success = Ok(EphemeralServer {
+            target,
+            has_test_service: config.has_test_service,
+            child,
+        });
+
+        // Try to connect every 100ms for 5s
+        // TODO(cretz): Some other way, e.g. via stdout, to know whether the
+        // server is up?
+        let client_options = ClientOptionsBuilder::default()
+            .identity("online_checker".to_owned())
+            .target_url(Url::parse(&target_url)?)
+            .client_name("online-checker".to_owned())
+            .client_version("0.1.0".to_owned())
+            .build()?;
+        for _ in 0..50 {
+            sleep(Duration::from_millis(100)).await;
+            if client_options
+                .connect_no_namespace(None, None)
```

Comment:
Is this enough? I don't remember what this does internally, does it issue a `getSystemInfo` call?

Reply:
Yes it does issue that call. It is enough.